theano/misc/pkl_utils.py

```diff
@@ -93,3 +113,217 @@ class CompatUnpickler(pickle._Unpickler):
+                (type(obj) is cuda_ndarray.CudaNdarray)):
+            if id(obj) not in self.seen:
+                def write_array(f):
+                    numpy.lib.format.write_array(f, numpy.asarray(obj))
+                name = self._resolve_name(obj)
+                zipadd(write_array, self.zip_file, name)
+                if type(obj) is cuda_ndarray.CudaNdarray:
+                    self.seen[id(obj)] = 'cuda_ndarray.{}'.format(name)
+                else:
+                    self.seen[id(obj)] = 'ndarray.{}'.format(name)
+            return self.seen[id(obj)]
+
+
+class PersistentSharedVariableID(PersistentNdarrayID):
+    """Persist the names of shared variable arrays in the zip file.
+
+    If a shared variable has a name, this name is used as the name of the
+    NPY file inside of the zip file. NumPy arrays that aren't matched to a
+    shared variable are persisted as usual (i.e. `array_0`, `array_1`,
+    etc.)
+
+    :param allow_unnamed: Allow shared variables without a name to be
+        persisted. Defaults to ``True``.
+    :type allow_unnamed: bool, optional
+
+    :param allow_duplicates: Allow multiple shared variables to have the same
+        name, in which case they will be numbered e.g. `x`, `x_2`, `x_3`, etc.
+        Defaults to ``True``.
+    :type allow_duplicates: bool, optional
+
+    :raises ValueError
+        If an unnamed shared variable is encountered and `allow_unnamed` is
+        ``False``, or if two shared variables have the same name, and
+        `allow_duplicates` is ``False``.
+
+    """
+    def __init__(self, zip_file, allow_unnamed=True, allow_duplicates=True):
+        super(PersistentSharedVariableID, self).__init__(zip_file)
+        self.name_counter = defaultdict(int)
+        self.ndarray_names = {}
+        self.allow_unnamed = allow_unnamed
+        self.allow_duplicates = allow_duplicates
+
+    def _resolve_name(self, obj):
+        if id(obj) in self.ndarray_names:
+            name = self.ndarray_names[id(obj)]
+            count = self.name_counter[name]
+            if count:
+                if not self.allow_duplicates:
+                    raise ValueError("multiple shared variables with the name "
+                                     "`{}` found".format(name))
+                name = '{}_{}'.format(name, count + 1)
+            self.name_counter[name] += 1
+            return name
+        return super(PersistentSharedVariableID, self)._resolve_name(obj)
+
+    def __call__(self, obj):
+        if isinstance(obj, SharedVariable):
+            if obj.name:
+                if obj.name == 'pkl':
+                    ValueError("can't pickle shared variable with name `pkl`")
+                self.ndarray_names[id(obj.container.storage[0])] = obj.name
+            elif not self.allow_unnamed:
+                raise ValueError("unnamed shared variable, {}".format(obj))
+        return super(PersistentSharedVariableID, self).__call__(obj)
```

Comment:
`super.__call__` will pass the `SharedVariable` to `PersistentNdarrayID`, which then returns `None`. This tells the pickler to simply pickle `SharedVariable` as usual. However, if `obj` is an `ndarray` or `CudaNdarray`, nothing in this child class method gets run, and the parent class does all the work.

The reason there is no `PersistentSharedVariableLoad` is because the names of the shared variables are pickled as part of the normal pickling process. We don't need a special class to load them. The only thing that needs to be loaded from the zip file is the actual `ndarray` or `CudaNdarray`, which is done by `PersistentNdarrayLoad`.